hypothesis-python/src/hypothesis/_strategies.py

```diff
@@ -775,6 +776,26 @@ def unique_by(x):
         for i, f in enumerate(unique_by):
             if not callable(f):
                 raise InvalidArgument("unique_by[%i]=%r is not a callable" % (i, f))
+        # Note that lazy strategies automatically unwrap when passed to a defines_strategy
+        # function.
+        if isinstance(elements, SampledFromStrategy):
```

Comment:
IIRC we went with the previous design because it also handled mapped and filtered strategies with the efficient path - and unfortunately the obvious way to handle that only works if we know that the function or predicate are pure functions, which they might not be...

It's kinda tedious, but if we're adding this special handling it would be nice if it could apply to `sampled_from(...).map(str).filter(bool)` too 😕